dali/python/nvidia/dali/_debug_mode.py

```diff
@@ -0,0 +1,332 @@
+            if res is None:
+                res = ()
+            elif not isinstance(res, tuple):
+                res = (res,)
+
+            # Transforming all variables to TensorLists.
+            return tuple([val.get() if isinstance(val, DataNodeDebug) else _Tensors.TensorListCPU(
+                np.tile(val, (self._max_batch_size, *[1]*np.array(val).ndim))) for val in res])
+
+    def feed_input(self, data_node, data, **kwargs):
+        """Pass data to an ExternalSource operator inside the pipeline.
+        
+        Refer to :meth:`Pipeline.feed_input() <nvidia.dali.Pipeline.feed_input>` for details."""
+        if not self._built:
+            raise RuntimeError("Pipeline must be built first.")
+
+        if data_node not in self._external_source_pipelines:
+            # Saving data, because pipeline hasn't been run yet.
+            if data_node not in self._feed_input_data:
+                self._feed_input_data[data_node] = []
+            self._feed_input_data[data_node].append((data, kwargs))
+        else:
+            self._external_source_pipelines[data_node].feed_input(data_node, data, **kwargs)
+
+    @staticmethod
+    def _classify_data(data):
+        # Returns tuple (use_external_source, device, unpacked data).
+
+        def is_primitive_type(x):
+            return isinstance(x, (int, float, bool, str))
+
+        if isinstance(data, list):
+            if any([is_primitive_type(d) for d in data]):
+                return False, 'cpu', data
+
+            device = None
+            data_list = []
+            for d in data:
+                _, cur_device, val = _PipelineDebug._classify_data(d)
+                if device == None:
+                    device = cur_device
+                data_list.append(val)
+            return True, device, data_list
+        else:
+            if is_primitive_type(data) or _types._is_numpy_array(data) or _types._is_mxnet_array(data) or isinstance(data, _Tensors.TensorCPU):
+                return False, 'cpu', data
+            if _types._is_torch_tensor(data):
+                return False, 'gpu' if data.is_cuda else 'cpu', data
+            if hasattr(data, '__cuda_array_interface__') or isinstance(data, _Tensors.TensorGPU):
+                return False, 'gpu', data
+            if isinstance(data, DataNodeDebug):
+                return True, data.device, data.get()
+            if isinstance(data, _Tensors.TensorListCPU):
+                return True, 'cpu', data
+            if isinstance(data, _Tensors.TensorListGPU):
+                return True, 'gpu', data
+
+        return False, 'cpu', data
+
+    def _create_subpipeline(self, op_wrapper, inputs, **kwargs):
+        """Creates pipeline wrapper around operator call (only once)."""
+
+        @_pipeline.pipeline_def(**self._subpipeline_kwargs)
+        def pipe():
+            inputs_preprocessed = []
+            kwargs_preprocessed = {}
+            for i, input in enumerate(inputs):
```

Comment:
Now I'm thinking about (probably very rare) case, where we recognize data as suitable for external source in one iter, and not in the other. Do you think we should capture this information here, and raise an error, somehow, when we detect that not all external inputs were fed/we tried to feed more than we had?
In first case it would probably block the pipeline, in the second it will most probably give us a bit cryptic error, that there is no external source of some `input_x` name.

Reply:
done, added checks and tests for it